File: core/src/com/alta_v2/rendering/common/dialog/DialogImpl.java
package com.alta_v2.rendering.common.dialog;

import com.alta_v2.rendering.Renderer;
import com.alta_v2.rendering.ScreenState;
import com.alta_v2.rendering.common.component.ComponentFactory;
import com.alta_v2.rendering.common.component.dialog.DialogComponent;
import com.alta_v2.rendering.common.component.dialog.TitleDialogStyleProvider;
import com.google.inject.Inject;
import lombok.extern.log4j.Log4j2;

@Log4j2
public class DialogImpl implements Renderer, TitleDialog {

    private final DialogComponent titleDialog;

    private float titleDialogDuration;
    private float currentTitleDialogDuration;

    @Inject
    public DialogImpl(ComponentFactory componentFactory, TitleDialogStyleProvider styleProvider) {
        titleDialog = componentFactory.createDialogComponent(
                styleProvider.createBoxInAnimation(), styleProvider.createBoxOutAnimation()
        );
    }

    @Override
    public void init(ScreenState state) {
        titleDialog.init(state);
    }

    @Override
    public void render(float delta, ScreenState state) {
        if (titleDialog.isVisible()) {
            currentTitleDialogDuration += delta;

            if (currentTitleDialogDuration >= titleDialogDuration) {
                hideTitle();
            }
        }

        titleDialog.render(delta, state);
    }

    @Override
    public void destroy() {
        titleDialog.destroy();
    }

    @Override
    public void showTitle(String text, float durationMs) {
        if (titleDialog.isVisible()) {
            log.warn("Title dialog already shown. Message '{}' won't shows", text);
            return;
        }
        titleDialog.show(text);
        titleDialogDuration = durationMs / 1000;
        currentTitleDialogDuration = 0;
    }

    @Override
    public void hideTitle() {
        if (titleDialog.isVisible()) {
            titleDialog.hide();
        }
    }
}

File: core/src/com/alta_v2/rendering/dialog/DialogImpl.kt
package com.alta_v2.rendering.dialog

import com.alta_v2.rendering.Renderer
import com.alta_v2.rendering.ScreenState
import com.alta_v2.rendering.common.component.dialog.DialogComponent
import com.google.inject.Inject
import mu.KotlinLogging

class DialogImpl @Inject constructor(dialogFactory: DialogComponentFactory) : Renderer, Dialog {

    private val log = KotlinLogging.logger {  }

    private val dialog = dialogFactory.createDialog()
    private val titleDialog = dialogFactory.createTitleDialog()

    private var titleDialogDuration = 0f
    private var currentTitleDialogDuration = 0f

    override fun init(state: ScreenState) {
        titleDialog.init(state)
        dialog.init(state)
    }

    override fun render(delta: Float, state: ScreenState) {
        if (titleDialog.visible) {
            currentTitleDialogDuration += delta
            if (currentTitleDialogDuration >= titleDialogDuration) {
                hideTitle()
            }
        }
        titleDialog.render(delta, state)

        dialog.render(delta, state)
    }

    override fun destroy() {
        titleDialog.destroy()
        dialog.destroy()
    }

    override fun showTitle(text: String, durationMs: Float) {
        if (titleDialog.visible) {
            log.warn("Title dialog already shown. Message '{}' won't shows", text)
            return
        }
        titleDialog.show(text, false)
        titleDialogDuration = durationMs / 1000
        currentTitleDialogDuration = 0f
    }

    override fun hideTitle() {
        hide(titleDialog)
    }

    override fun showDialog(text: String) = dialog.show(text, true)

    override fun hideDialog() = hide(dialog)

    override fun setDialogText(text: String) = dialog.setText(text, true)

    override fun isDialogVisible(): Boolean = dialog.visible

    override fun isDialogTextAnimationCompleted(): Boolean = dialog.isTextAnimationCompleted

    override fun completeDialogTextAnimationImmediately() = dialog.completeTextAnimationImmediately()

    private fun hide(d: DialogComponent) {
        if (d.visible) {
            d.hide()
        }
    }
}